### DISCUSSION TITLE
Enforce return value for validator

### DISCUSSION THREAD

**Discussion by PatrikHlobil** [pydantic/pydantic#2657] [Feature Request]
Dear mantainers of pydantic,

thank you for your work in this amazing library, I really enjoy working with it.

Nevertheless, I noticed one thing that I think could be improved in the way the validator works. Let me give you an easy example:

![image](https://user-images.githubusercontent.com/31643597/114349581-664a9500-9b68-11eb-8343-e9ad5bb98fe9.png)

Obviously, the validator works as expected an raises an Error when the first_name has more than 20 characters. When successfull however, the written validator has the subtle bug that it did return None because I forget the `return v`. I do not like this because:

1. "Explicit is better than implicit" and this behavior uses the implicit None-return type to set a value after validation, which I guess in 99% of the cases is not what the user wanted
2. It is a big root cause for potential bugs on validation. In our team, we had at least 2 instances where a forgotten `return v` changed the logic of the code that followed.

What I would propose as a possible alternative is the requirement that a validator method **has to have a return value** and otherwise raises an Error. This could be easily checked via the AST module or something similar. What do you think about this behavior? If you agree with me and would like to add such a feature, I of course would also help you with implementing this.

Best Patrik

**Comment by PrettyWood** [comment]
Hi @PatrikHlobil 
I agree that it's an easy way to missconfigure _pydantic_. A possible fix could be easy. Add a check when creating the validators
```py
def _prepare_validator(function: AnyCallable, allow_reuse: bool) -> classmethod:
    """
    Avoid validators with duplicated names since without this, validators can be overwritten silently
    which generally isn't the intended behaviour, don't run in ipython (see #312) or if allow_reuse is False.
    """
    import inspect

    f_cls = function if isinstance(function, classmethod) else classmethod(function)
    if not in_ipython() and not allow_reuse:
        ref = f_cls.__func__.__module__ + '.' + f_cls.__func__.__qualname__
        if ref in _FUNCS:
            raise ConfigError(f'duplicate validator function "{ref}"; if this is intended, set `allow_reuse=True`')
        if "return " not in inspect.getsource(function):
            raise ValueError(f'validator "{ref}" needs to return a value')
        _FUNCS.add(ref)
    return f_cls
```
Personally I would prefer `v` to be returned automatically without any explicit `return`. So validation would be more a "stop" or "modify" else continue...
Btw validation configuration may change in v2. See https://github.com/samuelcolvin/pydantic/issues/1984

**Comment by PatrikHlobil** [comment]
Hi @PrettyWood,

the principal idea is good. The implementation is however a little bit more difficult than shown in the example above. If you habe code like this it works:
```python
class Customer(BaseModel):
    name: str
    customer_type: str

    @validator("customer_type")
    def validate_customer_type(cls, v):
        if v == "premium":
            v = "premium user"

customer = Customer(name="Max", customer_type="premium")
```
![image](https://user-images.githubusercontent.com/31643597/116877725-985a8e80-ac1e-11eb-9a4f-c8a50b01a760.png)

However, for this code it does not work:
```python
class Customer(BaseModel):
    name: str
    customer_type: str

    @validator("customer_type")
    def validate_customer_type(cls, v):
        if v == "premium":
            return "premium user"
        print("All non-premium user will not have a return statement!")


customer = Customer(name="Max", customer_type="premium")
print(customer)

customer = Customer(name="Gwen", customer_type="standard")
print(customer)
```
![image](https://user-images.githubusercontent.com/31643597/116877999-fe471600-ac1e-11eb-9e85-18c13e99f88d.png)

Therefore, one would have to use AST or some other tool to evaluate all possible exit paths of the function and check that the function always exists with a return statement or an Exception.

What are your thoughts on this?

Best Patrik
